core/services/ocr2/plugins/ocr2keeper/evm/mercury_lookup.go

```diff
@@ -0,0 +1,343 @@
+			r.lggr.Debug("[MercuryLookup] callback reports upkeep not needed")
+			continue
+		}
+
+		// success!
+		upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_NONE
+		upkeepResults[i].State = types.Eligible
+		upkeepResults[i].PerformData = performData
+		r.lggr.Infof("[MercuryLookup] Success: %+v", upkeepResults[i])
+	}
+	// don't surface error to plugin bc MercuryLookup process should be self-contained.
+	return upkeepResults, nil
+}
+
+func (r *EvmRegistry) getUpkeepInfo(upkeepId *big.Int, opts *bind.CallOpts) (keeper_registry_wrapper2_0.UpkeepInfo, error) {
+	zero := common.Address{}
+	var err error
+	var upkeepInfo keeper_registry_wrapper2_0.UpkeepInfo
+	u, found := r.mercury.upkeepCache.Get(upkeepId.String())
+	if found {
+		upkeepInfo = u.(keeper_registry_wrapper2_0.UpkeepInfo)
+		r.lggr.Debugf("[MercuryLookup] cache hit UpkeepInfo: %+v\n", upkeepInfo)
+	} else {
+		upkeepInfo, err = r.registry.GetUpkeep(opts, upkeepId)
+		if err != nil {
+			return upkeepInfo, err
+		}
+		if upkeepInfo.Target == zero {
+			return upkeepInfo, errors.New("upkeepInfo should not be nil")
+		}
+		r.lggr.Debugf("[MercuryLookup] cache miss UpkeepInfo: %+v\n", upkeepInfo)
+		r.mercury.upkeepCache.Set(upkeepId.String(), upkeepInfo, cache.DefaultExpiration)
+	}
+	return upkeepInfo, nil
+}
+
+// decodeMercuryLookup decodes the revert error MercuryLookup(string feedLabel, string[] feeds, string feedLabel, uint256 query, byte[] extraData)
+func (r *EvmRegistry) decodeMercuryLookup(data []byte) (*MercuryLookup, error) {
+	e := r.mercury.abi.Errors["MercuryLookup"]
+	unpack, err := e.Unpack(data)
+	if err != nil {
+		return nil, fmt.Errorf("unpack error: %w", err)
+	}
+	errorParameters := unpack.([]interface{})
+
+	return &MercuryLookup{
+		feedLabel:  *abi.ConvertType(errorParameters[0], new(string)).(*string),
+		feeds:      *abi.ConvertType(errorParameters[1], new([]string)).(*[]string),
+		queryLabel: *abi.ConvertType(errorParameters[2], new(string)).(*string),
+		query:      *abi.ConvertType(errorParameters[3], new(*big.Int)).(**big.Int),
+		extraData:  *abi.ConvertType(errorParameters[4], new([]byte)).(*[]byte),
+	}, nil
+}
+
+// mercuryLookupCallback calls the callback(string[] memory chainlinkBlobHex, bytes memory extraData) specified by the
+// 4-byte selector from the revert. the return will match check telling us if the upkeep is needed and what the perform data is
+func (r *EvmRegistry) mercuryLookupCallback(ctx context.Context, mercuryLookup *MercuryLookup, values [][]byte, upkeepInfo keeper_registry_wrapper2_0.UpkeepInfo, opts *bind.CallOpts) (bool, []byte, error) {
+	payload, err := r.mercury.abi.Pack("mercuryCallback", values, mercuryLookup.extraData)
+	if err != nil {
+		return false, nil, fmt.Errorf("callback args pack error: %w", err)
+	}
+
+	//checkUpkeepGasLimit := uint32(200000) + uint32(6500000) + uint32(300000) + upkeepInfo.ExecuteGas
+	callbackMsg := ethereum.CallMsg{
+		From: r.addr,             // registry addr
```

Comment:
any reason for this to be via registry instead of zero address?

Reply:
updated